src/http.c

```diff
@@ -427,21 +424,41 @@ void
 http_encode(char *s, size_t len)
 {
 	char t[MAX_STRING];
+	char ch = 0;
 	unsigned i, j;
 
-	for (i = j = 0; s[i] && j < sizeof(t) - 1; i++, j++) {
-		t[j] = s[i];
-		if (s[i] <= 0x20 || s[i] >= 0x7f) {
-			/* Fix buffer overflow */
+	for (i = j = 0; i < len && j < sizeof(t) - 1 && (ch = s[i]); i++, j++) {
+		if (ch == 0x2a /* '*' */||
```

Comment:
too many cases, maybe should be moved to a function

Reply:
The cases are necessary, moving them inside a function can not resolve the complex, but cost 1 more function calling.
I suggest to keep it here, leaving enough comments for each case, will be much easier to maintain.

Reply:
There will be no extra cost, the compiler can optimize away the function.

Reply:
You can try yourself: https://godbolt.org/z/hc384Gn3d

There's no much difference between `-Os` and `-O2`.